### PR TITLE
Advanced setup

### PR DESCRIPTION
**Pull request by krokrob**
this adds a bunch of config for the data science bootcamp:
- check python packages installation and execution (including TensorFlow)
- check notebook server is up and running
- install and check `nbextensions`
- install and configure `gcloud`

### CODE REVIEW

**Comment by krokrob** [comment]
@gmanchon can you check the GCP section?
@brunolajoie can you check the VScode and pip sections?
@dmilon can you check the VScode section?

File: LINUX.md [hunk 1]
```diff
@@ -301,10 +315,10 @@ Open your terminal and go to your `dotfiles` project:
 
 ```bash
 cd ~/code/<YOUR_GITHUB_NICKNAME>/dotfiles
-stt # Open it in Sublime Text
+vsc # Open it in VScode
 ```
 
-In Sublime Text, open the `zshrc` file. Replace its content with the [newest version](https://raw.githubusercontent.com/lewagon/dotfiles/master/zshrc) of that file that we provide. Save to disk.
+In VScode, open the `zshrc` file. Replace its content with the [newest version](https://raw.githubusercontent.com/lewagon/dotfiles/master/zshrc) of that file that we provide. Save to disk.
```
Comment (dmilon) [inline LINUX.md:321]:
```suggestion
In VS Code, open the `zshrc` file. Replace its content with the [newest version](https://raw.githubusercontent.com/lewagon/dotfiles/master/zshrc) of that file that we provide. Save to disk.
```

File: LINUX.md [hunk 2]
```diff
@@ -301,10 +315,10 @@ Open your terminal and go to your `dotfiles` project:
 
 ```bash
 cd ~/code/<YOUR_GITHUB_NICKNAME>/dotfiles
-stt # Open it in Sublime Text
+vsc # Open it in VScode
```
Comment (dmilon) [inline LINUX.md:318]: I think that in the data program on kitt, the text editor command is still `code .`. Introducing an alias would require to update the program and check in the content there is no more `code .`. IMHO, it doesn't worth it for students. Teachers can do as they please of course.

File: LINUX.md [hunk 3]
```diff
@@ -375,10 +389,10 @@ Please now **quit** all your opened terminal windows.
 &nbsp;
 
 
-Let us open the `~/.zshrc` profile file in Sublime Text and change slightly its content:
+Let us open the `~/.zshrc` profile file in Visual Studio Code and change slightly its content:
```
Comment (dmilon) [inline LINUX.md:392]:
```suggestion
Let us open the `~/.zshrc` profile file in VS Code and change slightly its content:
```

File: LINUX.md [hunk 4]
```diff
@@ -406,7 +420,7 @@ plugins=(gitfast last-working-dir common-aliases sublime zsh-syntax-highlighting
 &nbsp;
 
 
-&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close Sublime Text.
+&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close Visual Studio Code.
```
Comment (dmilon) [inline LINUX.md:423]:
```suggestion
&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close VS Code.
```

File: WINDOWS.md [hunk 5]
```diff
@@ -1030,10 +1043,10 @@ Open your terminal and go to your `dotfiles` project:
 
 ```bash
 cd ~/code/<YOUR_GITHUB_NICKNAME>/dotfiles
-stt # Open it in Sublime Text
+vsc # Open it in VScode
```
Comment (dmilon) [inline WINDOWS.md:1046]:
```suggestion
code . # Open it in VS Code
```

File: LINUX.md (hunk 2 above)
Comment (dmilon) [reply LINUX.md:318]:
```suggestion
code . # Open it in VS Code
```

File: LINUX.md [hunk 6]
```diff
@@ -500,16 +514,101 @@ pip install requests bs4
...
+
+Find the `jupyter` config directory:
+``` bash
+JUPYTER_CONFIG_DIR=$(jupyter --config-dir)
+```
+
+Open `custom/custom.js` in the config directory:
+```bash
+cd $JUPYTER_CONFIG_DIR
+vsc custom/custom.js
```
Comment (dmilon) [inline LINUX.md:564]:
```suggestion
code custom/custom.js
```

File: LINUX.md [hunk 7]
```diff
@@ -500,16 +514,101 @@ pip install requests bs4
...
+        var cells = IPython.notebook.get_cells();
+        for(var cl=0; cl< cells.length ; cl++){
+            cells[cl].code_mirror.setOption('keyMap', 'sublime');
+        }
+        // }, 1000)// uncomment  line to fake race condition
+    }
+);
+```
+
+You can close VScode.
```
Comment (dmilon) [inline LINUX.md:581]:
```suggestion
You can close VS Code.
```

File: WINDOWS.md [hunk 8]
```diff
@@ -1255,7 +1274,7 @@ jupyter notebook --generate-config
 We will now edit the generated Jupyter configuration file:
 
 ``` bash
-code $HOME/.jupyter/jupyter_notebook_config.py
+vsc $HOME/.jupyter/jupyter_notebook_config.py
```
Comment (dmilon) [inline WINDOWS.md:1277]:
```suggestion
code $HOME/.jupyter/jupyter_notebook_config.py
```

File: WINDOWS.md [hunk 9]
```diff
@@ -1285,6 +1304,85 @@ If it is not the case, please call a TA.
...
+
+Find the `jupyter` config directory:
+``` bash
+JUPYTER_CONFIG_DIR=$(jupyter --config-dir)
+```
+
+Open `custom/custom.js` in the config directory:
+```bash
+cd $JUPYTER_CONFIG_DIR
+vsc custom/custom.js
```
Comment (dmilon) [inline WINDOWS.md:1336]:
```suggestion
code custom/custom.js
```

File: _partials/dotfiles.md [hunk 10]
```diff
@@ -10,10 +10,10 @@ Open your terminal and go to your `dotfiles` project:
 
 ```bash
 cd ~/code/<YOUR_GITHUB_NICKNAME>/dotfiles
-stt # Open it in Sublime Text
+vsc # Open it in VScode
```
Comment (dmilon) [inline _partials/dotfiles.md:13]:
```suggestion
code . # Open it in VS Code
```

File: _partials/nbextensions.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+Find the `jupyter` config directory:
+``` bash
+JUPYTER_CONFIG_DIR=$(jupyter --config-dir)
+```
+
+Open `custom/custom.js` in the config directory:
+```bash
+cd $JUPYTER_CONFIG_DIR
+vsc custom/custom.js
```
Comment (dmilon) [inline _partials/nbextensions.md:30]:
```suggestion
code custom/custom.js
```

File: _partials/nbextensions.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        var cells = IPython.notebook.get_cells();
+        for(var cl=0; cl< cells.length ; cl++){
+            cells[cl].code_mirror.setOption('keyMap', 'sublime');
+        }
+        // }, 1000)// uncomment  line to fake race condition
+    }
+);
+```
+
+You can close VScode.
```
Comment (dmilon) [inline _partials/nbextensions.md:47]:
```suggestion
You can close VS Code.
```

File: macOS.md [hunk 13]
```diff
@@ -565,7 +572,7 @@ plugins=(gitfast last-working-dir common-aliases sublime zsh-syntax-highlighting
 &nbsp;
 
 
-&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close Sublime Text.
+&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close Visual Studio Code.
```
Comment (dmilon) [inline macOS.md:575]:
```suggestion
&nbsp;&nbsp;&nbsp; :white_check_mark: Save the `.zshrc` file with `Ctrl` + `S` and close VS Code.
```

**Review by dmilon — CHANGES_REQUESTED** [review]
:+1: :pray: I've checked the VS Code part @krokrob. 

I think aliasing `code .` to `vsc` doesn't worth it as `code .` is pretty straightforward already and changing this here would require to edit the program text editor and check that no more `code .` remains in the content. 

Other comments are just a proposal to standardize the name of the text editor: let's call it "VS Code" as it's most common abbreviation found on the web.

**Comment by krokrob** [comment]
Tkanks @dmilon 🙏

File: _partials/nbextensions.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+```bash
+# install nbextensions
+pip install jupyter_contrib_nbextensions
+pip install jupyter_nbextensions_configurator
+jupyter contrib nbextension install --user
+jupyter nbextension enable toc2/main
+jupyter nbextension enable collapsible_headings/main
+jupyter nbextension enable spellchecker/main
+jupyter nbextension enable code_prettify/
+jupyter nbextension enable autopep8
```
Comment (brunolajoie) [inline _partials/nbextensions.md:14]: autopep8 and code_prettify seems redundant. Why install both?

Comment (krokrob) [reply _partials/nbextensions.md:14]: I'll remove `code_prettify`

File: _partials/vscode_setup.md [hunk 15]
```diff
@@ -1,18 +1,12 @@
...
-- In VS Code, click on `install`
+Let's gain time now and add other extensions that will be helpful during your Bootcamp:
 
-**Extensions**:
-- [Sublime Text Keymap](https://marketplace.visualstudio.com/items?itemName=ms-vscode.sublime-keybindings)
-- [Python](https://marketplace.visualstudio.com/items?itemName=ms-python.python)
-- [Jupyter](https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter)
-- [Python Indent](https://marketplace.visualstudio.com/items?itemName=KevinRose.vsc-python-indent)
+```bash
+<CODE_EDITOR_CMD> --install-extension ms-vscode.sublime-keybindings ms-python.python ms-toolsai.jupyter KevinRose.vsc-python-indent MS-vsliveshare.vsliveshare
```
[the remark this answers is not in the document]
Comment (brunolajoie) [reply _partials/vscode_setup.md:6]: Also, the sublime-keybindings extention listed here is redundant with the paragraph just before that asks to install sublime manually (see screenshot below)
image

File: _partials/vscode_setup.md [hunk 16]
```diff
@@ -1,18 +1,12 @@
...
+```
 
-### VS Code Settings
+### <CODE_EDITOR> Settings
 
 - On macOS, press `Cmd` + `,` on your keyboard (or `Ctrl` + `,` on Windows or Linux) to open the settings
 - In the search bar, type `emmet`
```
Comment (brunolajoie) [inline _partials/vscode_setup.md:12]: There is one setting I have to make on my machine to make VS code works with pyenv:

```bash
{
  "python.pythonPath": "~/.pyenv/shims/python",
}
```

In addition, we should also use exactly the same auto-formatter between jupyter and vscode: Therefore if you told them to install the jupyter extention `code_prettify`, you have to specify `yapf` for VS code too. Remember that `code_prettify` == `yapf` == `pep8 enforcement + nicer prettyfication` (by default) 
```bash
{
  "python.formatting.provider": "yapf", 
}
```

Finally, optionally you may want to add this so they don't clutter their left navbar

```bash
{
  "files.exclude": {
      "**/.pytest_cache": true,
      "**/.ipynb_checkpoints": true,
      "**/__pycache__": true,
      "**/.egg-info": true,
    },
}
```

File: LINUX.md [hunk 17]
```diff
@@ -500,16 +508,100 @@ pip install requests bs4
 Finally, more Data Science packages:
 
 ```bash
-pip install jupyterlab pandas matplotlib seaborn plotly scikit-learn
+pip install jupyterlab pandas matplotlib seaborn plotly scikit-learn tensorflow
```
Comment (brunolajoie) [inline LINUX.md:511]: I recommand installing more of them so as to detect potential zlma errors & co
```bash
autopep8 yapf dtale nbconvert
requests
statsmodels xgboost pandas-profiling
```

**Review by brunolajoie — CHANGES_REQUESTED** [review]

**Pull request merged**